### PR TITLE
Add an extract_zipkin_attrs_from_headers() helper

### PR DESCRIPTION
**Pull request by dbishop**
It turns out that extracting B3 headers correctly is harder than it
sounds.  This helper function implements the standard as specified at:
https://github.com/openzipkin/b3-propagation

You can pass a dict or dict-like object to the new
extract_zipkin_attrs_from_headers() function, and get back a
ZipkinAttrs instance you can trust.

In some cases, a sample_rate float and a use_128bit_trace_id flag are
required, so they appear in the function signature as well.

This should take care of single-header extraction, which is part of
Issue #98.

This may help with Issue #73 (missing or empty X-B3-Sampled).

### CODE REVIEW

File: py_zipkin/zipkin.py [hunk 1]
```diff
@@ -639,6 +640,12 @@ def __init__(self, *args, **kwargs):
         super(zipkin_server_span, self).__init__(*args, **kwargs)
 
 
+def choose_sampling(sample_rate):
```
Comment (drolando) [inline py_zipkin/zipkin.py]: nit: maybe call this `should_sample`? choose_sampling gives me the idea that it'll return the sample rate.

Also, I'd add a leading `_` to this function since it doesn't seem something that's useful outside of this code.

File: py_zipkin/zipkin.py [hunk 2]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
         'X-B3-Flags': '0',
         'X-B3-Sampled': '1' if zipkin_attrs.is_sampled else '0',
     }
+
+
+def is_valid_trace_id(candidate):
```
Comment (drolando) [inline py_zipkin/zipkin.py]: Please remove validation or at least make it optional and disabled by default. Even though the spec says that ids should be 16 or 32 bytes long, the zipkin UI and api accepts also shorter ids and pads them with leading 0s.

This has unfortunately led to systems in our infrastructure using shorter ids, which wouldn't pass this validation.

Your call, but I'd just get rid of the validation which will also simplify your tests a lot

File: py_zipkin/zipkin.py [hunk 3]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
...
+
+
+def _extract_zipkin_attrs_from_single_header(b3_header, sample_rate,
+                                             use_128bit_trace_id):
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    # where the last two fields are optional.
+    bits = b3_header.split('-')
+    if len(bits) == 1:
+        # See if it's a valid encoding of (only) sample decision
+        if bits[0] == 'false' or bits[0] == '0':
```
Comment (drolando) [inline py_zipkin/zipkin.py]: I don't think `false` is a valid sample header. Reading https://github.com/openzipkin/b3-propagation/blob/master/RATIONALE.md it seems like it could be either `0`, `1` or `d` but the description of this case is very unclear.
cc @adriancole

File: py_zipkin/zipkin.py [hunk 4]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
...
+def is_valid_span_id(candidate):
+    return candidate and re.match(r'[0-9a-f]{16}$', candidate)
+
+
+def _extract_zipkin_attrs_from_single_header(b3_header, sample_rate,
+                                             use_128bit_trace_id):
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    # where the last two fields are optional.
+    bits = b3_header.split('-')
+    if len(bits) == 1:
```
Comment (drolando) [inline py_zipkin/zipkin.py]: Add a comment that there could also be just 1 field and that's the sampling decision. Otherwise this if seems in contradiction with the above comment that says only the last 2 fields are optional

File: py_zipkin/zipkin.py [hunk 5]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
...
+            use_128bit_trace_id=use_128bit_trace_id)
+    if not is_valid_trace_id(bits[0]) or not is_valid_span_id(bits[1]):
+        return None
+    if len(bits) == 4:
+        if not is_valid_span_id(bits[3]):
+            return None
+        parent_span_id = bits[3]
+    else:
+        parent_span_id = None
+    if len(bits) >= 3 and bits[2]:
```
Comment (drolando) [inline py_zipkin/zipkin.py]: add an empty line before this

File: py_zipkin/zipkin.py [hunk 6]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
...
+        elif bits[2] == '0' or bits[2] == 'false':
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            return None
+    else:
+        # sample flag missing; means "Defer" and we're responsible for
+        # rolling fresh dice
+        is_sampled = choose_sampling(sample_rate)
+    return ZipkinAttrs(bits[0], bits[1], parent_span_id,
```
Comment (drolando) [inline py_zipkin/zipkin.py]: one arg per line

File: py_zipkin/zipkin.py [hunk 7]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
...
+    The input headers can be any dict-like container that supports "in"
+    membership test and a .get() method that accepts a default value.
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        return _extract_zipkin_attrs_from_single_header(
+            headers['b3'], sample_rate, use_128bit_trace_id)
+
+    # No single header, let's get old-school
```
Comment (drolando) [inline py_zipkin/zipkin.py]: since you have an helper function for `_extract_zipkin_attrs_from_single_header` I'd add one also for the regular headers

File: py_zipkin/zipkin.py [hunk 8]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
...
+        # We have 1==Accept, 0==Deny
+        # Also, "While you shouldn't encode X-B3-Sampled as true or false, a
+        #        lenient implementation may accept them."
+        if sampled_str == '1' or sampled_str == 'true':
+            is_sampled = True
+        elif sampled_str == '0' or sampled_str == 'false':
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            return None
```
Comment (drolando) [inline py_zipkin/zipkin.py]: Maybe log a warning in these cases where the headers are not valid?

File: tests/zipkin_test.py [hunk 9]
```diff
@@ -885,3 +885,464 @@ def test_create_headers_for_new_span_custom_tracer(gen_mock):
...
         tracer=tracer,
     )
+
+
+@mock.patch('py_zipkin.zipkin.create_attrs_for_span')
+def test_extract_zipkin_attrs_from_headers_single_sample_only(mock_create_attrs):
+    mock_create_attrs.return_value = 'xyz'  # just a sentinel
+    # While you shouldn't encode X-B3-Sampled as true or false, a lenient
+    # implementation may accept them
+    for no_sample_string in ('0', 'false'):
```
Comment (drolando) [inline tests/zipkin_test.py]: use pytest.mark.parameterize instead

File: tests/zipkin_test.py [hunk 10]
```diff
@@ -885,3 +885,464 @@ def test_create_headers_for_new_span_custom_tracer(gen_mock):
     assert expected == zipkin.create_http_headers_for_new_span(
         tracer=tracer,
     )
+
+
+@mock.patch('py_zipkin.zipkin.create_attrs_for_span')
+def test_extract_zipkin_attrs_from_headers_single_sample_only(mock_create_attrs):
+    mock_create_attrs.return_value = 'xyz'  # just a sentinel
```
Comment (drolando) [inline tests/zipkin_test.py]: I'd get rid of this and let the function return a real ZipkinAttr.

In the cases where you don't care about the exact value of the ids (like when you only pass the sample flag), just check that specific value.

This will also get rid of the extra `assert mock_create_attrs.mock_calls` making the tests easier to read

File: py_zipkin/zipkin.py [hunk 11]
```diff
@@ -709,3 +713,125 @@ def create_http_headers_for_new_span(context_stack=None, tracer=None):
         'X-B3-Flags': '0',
         'X-B3-Sampled': '1' if zipkin_attrs.is_sampled else '0',
     }
+
+
```
Comment (drolando) [inline py_zipkin/zipkin.py]: This file is getting very long and these functions don't really depend on anything else defined in here. What about creating a new file, something like request_helpers.py or so

Comment (dbishop) [reply py_zipkin/zipkin.py]: Will do!  I'm wrangling the structure around, then I'll start in on all the other (great!) suggestions.

File: py_zipkin/zipkin.py (hunk 3 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: Makes sense that single `b3` header would not have the legacy baggage of the old `X-B3-Sample` header.

Removing that leniency.

File: py_zipkin/zipkin.py (hunk 2 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: No prob, ripping it out :)

File: py_zipkin/zipkin.py (hunk 1 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: Done

File: py_zipkin/zipkin.py (hunk 4 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: Done

File: py_zipkin/zipkin.py (hunk 5 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: Done (including additional cleanup & commenting)

File: py_zipkin/zipkin.py (hunk 6 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: Done

File: py_zipkin/zipkin.py (hunk 7 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: I did, but it's literally just repeating the function signature of the enclosing function.

```
    return _extract_zipkin_attrs_from_multiple_headers(
        headers, sample_rate, use_128bit_trace_id)
```

I could extract the values from headers and then call a helper function, but then it'd need to take 5 args.  Not sure that helps readability, either.

File: py_zipkin/zipkin.py (hunk 8 above)
Comment (dbishop) [reply py_zipkin/zipkin.py]: Done (throughout).  I picked `log.warning(...)` let me know if another level is more appropriate.

File: tests/zipkin_test.py (hunk 9 above)
Comment (dbishop) [reply tests/zipkin_test.py]: Done

File: tests/zipkin_test.py (hunk 10 above)
Comment (dbishop) [reply tests/zipkin_test.py]: Done

File: tests/request_helpers_test.py [hunk 12]
```diff
@@ -0,0 +1,408 @@
+# -*- coding: utf-8 -*-
+import mock
+import pytest
+
+import py_zipkin.request_helpers as request_helpers
```
Comment (drolando) [inline tests/request_helpers_test.py]: why not `from py_zipkin import request_helpers`?

File: tests/request_helpers_test.py [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+from py_zipkin.request_helpers import ZipkinAttrs
+
+
+def test_extract_zipkin_attrs_from_headers_single_sample_only():
+    # While you shouldn't encode X-B3-Sampled as true or false, a lenient
```
Comment (drolando) [inline tests/request_helpers_test.py]: this comment is now wrong

File: tests/request_helpers_test.py [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+    assert ZipkinAttrs(
+        trace_id, span_id, None, '0', False,
+    ) == request_helpers.extract_zipkin_attrs_from_headers({
+        'X-B3-TraceId': trace_id,
+        'X-B3-SpanId': span_id,
+    }, sample_rate=88.2)
+    assert [
+        mock.call()
+    ] == mock_random.mock_calls
+    mock_random.reset_mock()
```
Comment (drolando) [inline tests/request_helpers_test.py]: why do you care? The response is correct and I don't think these extra asserts add much value and make the tests even longer

File: tests/request_helpers_test.py [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+        assert ZipkinAttrs(
+            trace_id, span_id, None, '0', True,
+        ) == request_helpers.extract_zipkin_attrs_from_headers({
+            'X-B3-TraceId': trace_id,
+            'X-B3-SpanId': span_id,
+            'X-B3-Sampled': yes_trace_str,
+        }, sample_rate=88.2)
+        assert [] == mock_random.mock_calls
+
+    # non-defer debug-trace, no parent span
```
Comment (drolando) [inline tests/request_helpers_test.py]: there's no `d` in here

File: tests/request_helpers_test.py [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+        'X-B3-ParentSpanId': parent_span_id,
+    }, sample_rate=88.2)
+    assert [] == mock_random.mock_calls
+
+
+@mock.patch('py_zipkin.util.random.random')
+def test_extract_zipkin_attrs_from_headers_single(mock_random):
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    # where the last two fields are optional.
+
```
Comment (drolando) [inline tests/request_helpers_test.py]: Do we actually need all these tests to cover all the cases? There's a lot of them and they seem to repeat themselves.

I think the repeating part stems from the fat that we're basically implementing the same logic twice. Could we refactor away the common logic in request_helpers.py so that we only have to test it once? The only difference between the 2 afaict is whether all the fields are in a single header or in multiple ones.

So we could just have the logic to parse the headers t be separate and then call the same underlying function to generate the ZipkinAttrs

File: tests/request_helpers_test.py (hunk 12 above)
Comment (dbishop) [reply tests/request_helpers_test.py]: Whoops, just a brain fart; fixing

File: tests/request_helpers_test.py (hunk 13 above)
Comment (dbishop) [reply tests/request_helpers_test.py]: Fixed

File: tests/request_helpers_test.py (hunk 14 above)
Comment (dbishop) [reply tests/request_helpers_test.py]: Fair point; removing.

File: tests/request_helpers_test.py (hunk 15 above)
Comment (dbishop) [reply tests/request_helpers_test.py]: I went back and double-checked https://github.com/openzipkin/b3-propagation and the "d" thing is only for single-header AFAICT.

For multi-header, it says only (no mention of "d" for Sampling State nor Debug Flag):
```
Sampling State

An accept sampling decision is encoded as X-B3-Sampled: 1 and a deny as X-B3-Sampled: 0. Absent means defer the decision to the receiver of this header. For example, a Sampled header might look like: X-B3-Sampled: 1.

Note: Before this specification was written, some tracers propagated X-B3-Sampled as true or false as opposed to 1 or 0. While you shouldn't encode X-B3-Sampled as true or false, a lenient implementation may accept them.
Debug Flag

Debug is encoded as X-B3-Flags: 1. Debug implies an accept decision, so don't also send the X-B3-Sampled header.
```

File: tests/request_helpers_test.py (hunk 16 above)
Comment (dbishop) [reply tests/request_helpers_test.py]: The specification is a little loose and I kept running into wacky under-specified edge cases.  I think the volume of tests is justified by that.

Personally, I'd rather have overly-exhaustive tests than untested edge cases.

Because of the spec, it's not _quite_ apples to apples.  The single-header form is newer and drops some backwards-compat, the "d" thing is only in single-header, "flags" isn't a thing in single-header.

I guess those impedance mismatches could be smoothed over, but because of the spec, you'd really want the tests like they are to make sure _that_ code is right and each side of the spec (multi-header/single-header) are correct.

Refactoring the implementation to have fewer lines seems reasonable for reasons like clarity, DRY, less bug surface area, etc., but I'm not sure having shorter tests is a great justification.

Let me see what the refactoring looks like/

Comment (dbishop) [reply tests/request_helpers_test.py]: @drolando That was a great idea, and I think the code cleaned up nicely.  Please see latest commit.

I still the the tests are good, as they stand now, for the reasons I mentioned earlier.

Comment (adriancole) [reply tests/request_helpers_test.py]: ps only one flag value was ever defined, 1, so not sure how much under-specifying we'd need to do to clarify d is same as presence of that flag.. can we fix the spec somehow?

File: py_zipkin/request_helpers.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
+    # non-empty, but invalid string.
+    if parsed['sampled_str'] == '':
+        parsed['sampled_str'] = 'INVALID'
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: why not use exception handling or just quit out.. this is corrupt.

File: py_zipkin/request_helpers.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    parsed['sample_decision_only'] = (
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: if this is actually needed, it seems more like a utility than something needed in a data structure.. that or maybe it is conventional in python to do things like this.

File: py_zipkin/request_helpers.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        all(parsed[k] is None for k in ('trace_id', 'span_id', 'parent_span_id'))
+        and parsed['sampled_str'])
+    return parsed
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: why does extracting use a sample rate? The result of extracting headers is just the data extracted.. turning the headers into a new context sure.. but why are the concerns conflated here?

File: py_zipkin/request_helpers.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        parsed = _parse_single_header(headers['b3'])
+    else:
+        parsed = _parse_multi_header(headers)
+
+    # Handle the lone-sampling-decision case:
+    if parsed['sample_decision_only']:
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: traceId and spanId are required unless it is sampling decision only. you can just check for traceId after ensuring your above implementations don't accidentally allow an invalid/corrupt extraction to return

File: py_zipkin/request_helpers.py [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        # See if it's a valid encoding of (only) sample decision
+        if parsed['sampled_str'] in ('0', 'false'):
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: it is more common to have separate keys.. one for sampled and one for debug, prioritizing debug as it implies sampled. In cases where formats permit old handling (only b3 multi would allow false), you'd keep parsing coupled to that, vs later here. That means the parse result would have a tri-state boolean for sampled. ex the 'sampled' key doesn't exist, or it does and it is true or false (or some explicit encoding of that ex 1 or 0)

File: py_zipkin/request_helpers.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+            sample_rate = 0.0
+        elif parsed['sampled_str'] in ('1', 'true', 'd'):
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: debug wins always, so this moves to the top.. it is also unintuitive to handle false before true..

File: py_zipkin/request_helpers.py [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+            sample_rate = 100.0
+        else:
+            # Nonsense gets a None
+            log.warning('Ignoring invalid sampling decision %r; '
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: logs like this should occur at parsing as for example, b3 single has a smaller amount of valid options than multi. Once you verify you can't return nonsense from your own impl of b3 single and multi, you'd have no reason to double check here as the implementation is sealed to this file.

File: py_zipkin/request_helpers.py [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+            return None
+        attrs = create_attrs_for_span(
+            sample_rate=sample_rate,
+            use_128bit_trace_id=use_128bit_trace_id,
+            flags='1' if parsed['sampled_str'] == 'd' else '0',
+        )
+        return attrs
+
+    # Enforce requirement of having trace & span IDs
+    if not parsed['trace_id'] or not parsed['span_id']:
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: again this should happen in the actual format extraction. this could be a function used in both places after you've handled the sampled-only branch.

File: py_zipkin/request_helpers.py [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+        log.warning('Ignoring invalid or missing TraceId & SpanId:'
+                    '%r, %r', parsed['trace_id'], parsed['span_id'])
+        return None
+
+    # Handle sampling decision
+    if parsed['sampled_str']:
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: not sure why we are doing this again.. similar comment to elsewhere.. be tight where the actual parsing occurs (ex you can log the input there if needed). then you don't need to do it again later.

File: py_zipkin/request_helpers.py [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+            is_sampled = False
+        else:
+            # Not valid according to spec
+            log.warning('Invalid SamplingState %r; requires '
+                        '"0", "1", or "d".', parsed['sampled_str'])
+            return None
+    else:
+        # sample flag missing; means "Defer" and we're responsible for
+        # rolling fresh dice
+        is_sampled = _should_sample(sample_rate)
```
Comment (adriancole) [inline py_zipkin/request_helpers.py]: I'd do this externally, especially as sampling like this is only one form of sampling. ex later it could be done via http parameters. having the sample rate code here is unnecessarily coupled.

File: py_zipkin/util.py [hunk 27]
```diff
@@ -58,3 +58,11 @@ def signed_int_to_unsigned_hex(signed_int):
     if hex_string.endswith('L'):
         return hex_string[:-1]
     return hex_string
+
+
+def _should_sample(sample_rate):
```
Comment (adriancole) [inline py_zipkin/util.py]: similar to above, this shouldn't be a part of a change around parsing..

File: tests/request_helpers_test.py (hunk 16 above)
Comment (dbishop) [reply tests/request_helpers_test.py]: (I misspoke earlier and said `"flags" isn't a thing in single-header`; I meant `"flags" isn't a thing in multi-header`)

I'm not sure I understand what you're saying.  I think the spec and the code of this PR are clear in that:
1) single-header b3 has a tri-state `SamplingState` segment with "1", "0", and "d" defined where "d" means "debug" and is like a "1" with a bit of additional semantic import.
2) multi-header has an ill-fated `X-B3-Flags` header where `1` means `d` (i.e. debug), and apparently no other flags are defined (not even `0` for "there are no flags" or "debug is not active"?)

As far as fixing the spec, I can only assume other folks have already coded to it and the cat is out of the bag?  Like the true/false thing: I decided to go ahead and be lenient in both encodings here wrt `true`/`false` values of the sampling decision; after the refactoring, either both encodings accept them or neither do.  I decided that being lenient would be better.

File: py_zipkin/request_helpers.py [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+def _parse_multi_header(headers):
+    """
+    Parse out and return the data necessary for generating ZipkinAttrs.
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
```
[the remark this answers is not in the document]
Comment (dbishop) [reply py_zipkin/request_helpers.py]: The difference is that when only a sampling decision is specified, trace_id/span_id need to be generated (the call to `create_attrs_for_span()`).

Do you mean the logic to determine if we're in that state (sample decision only) should live in `extract_zipkin_attrs_from_headers()` instead of the parser helpers?

File: py_zipkin/request_helpers.py [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+        'span_id': headers.get('X-B3-SpanId', None),
+        'parent_span_id': headers.get('X-B3-ParentSpanId', None),
+        'sampled_str': headers.get('X-B3-Sampled', None),
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
```
[the remark this answers is not in the document]
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Just because you and I both guessed right doesn't make the spec unambiguous.  Is it necessary for specifications to be unambiguous? No, but it simplifies implementing them. I'd expect any disambiguation you could do in the spec would help anyone trying to implement it, not just me.

File: py_zipkin/request_helpers.py (hunk 18 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Oh, I wouldn't blame the language for my infelicities ;)

This is the second time you mentioned something about this key, so I'll see about moving that logic out of the parsing.

File: py_zipkin/request_helpers.py (hunk 17 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: ...because I didn't think of it.  But that's a great idea!  I'll have the parsers use exceptions to bail on invalid input and that might simplify the common post-parsing logic.

File: py_zipkin/request_helpers.py (hunk 19 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Well, this is my call-site, which I'm pretty happy with:
```
# Note: this is a monkey-patch for eventlet.wsgi.HttpProtocol.handle_one_response
# Also api.sample_rate_pct is a global for the process set via config file
def _patched_handle_one_response(self):
    zipkin_attrs = api.extract_zipkin_attrs_from_headers(
        self.headers, sample_rate=api.sample_rate_pct, use_128bit_trace_id=True)
    ...
    with api.ezipkin_server_span(
        service_name=api.default_service_name(),
        span_name=self.command,
        zipkin_attrs=zipkin_attrs,
        sample_rate=None if zipkin_attrs else api.sample_rate_pct,
...
    ) as zipkin_span:
        ...
```
Is there something undesirable there that I'm missing?

File: py_zipkin/request_helpers.py (hunk 20 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Right!  This is getting back to the raise-exception-on-bad-input in the parsers things.  Love it!

File: py_zipkin/request_helpers.py (hunk 21 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: I don't follow your first sentence.  But I think you're saying you don't want `true`/`false` accepted in the single-header case?

File: py_zipkin/request_helpers.py (hunk 22 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: I haven't found that to be the case, myself, but it's easy enough for me to change the order.

File: py_zipkin/request_helpers.py (hunk 23 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Okay, I think this is getting back to the theme of validating the input in the parsers, including logging bad output, and using exceptions for flow control.  I dig it!

File: py_zipkin/request_helpers.py (hunk 24 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Makes sense.

File: py_zipkin/request_helpers.py (hunk 26 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Well, the `ZipkinAttrs` are a namedtuple, which is a tuple, which is immutable.  If a `ZipkinAttrs` gets out of here with a `is_sampled` field that just needs to get filled in immediately, a whole new ZipkinAttrs would need to be created.

That just seems a bit byzantine to me...

I'll refer you to my example call-site above in another comment and ask again how that could be better?  I'm pretty happy with it.

File: py_zipkin/request_helpers.py (hunk 25 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: I'm sensing a theme about handing bad input in the parsing, perhaps with exceptions for flow control, simplifying the common code post-parsing...

File: py_zipkin/util.py (hunk 27 above)
Comment (dbishop) [reply py_zipkin/util.py]: similar to above, I think it's pretty good and don't understand the rationale for not having the `ZipkinAttrs` that come out being ready to feed into `zipkin_span()`.

**Comment by dbishop** [comment]
@adriancole thanks for the prompt review, I think validating in the parsers, moving the logging there, and using exceptions for flow control will simplify things nicely!

I'm working up a new commit for that now.  👍

File: py_zipkin/request_helpers.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (adriancole) [reply py_zipkin/request_helpers.py]: I mean it is surprising to have a field that is basically caching a result of a function you can do anytime after the parse. it being present in the data structure is distracting.

File: py_zipkin/request_helpers.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (adriancole) [reply py_zipkin/request_helpers.py]: I think it is overly pendantic, but I'll update the spec to explicitly say invalid data is not to be interpreted the same as if you never sent the header.

[the remark this answers is not in the document]
Comment (adriancole) [reply py_zipkin/request_helpers.py]: https://github.com/openzipkin/b3-propagation/pull/40

File: py_zipkin/request_helpers.py (hunk 19 above)
Comment (adriancole) [reply py_zipkin/request_helpers.py]: in other codebases, we reuse code like you've written in scenarios that don't assume a specific sampling rate implementation. To make things more flexible, we compose the feature to extract the headers with whatever it is that needs it (ex applying a rate). By keeping these things separate, we can also clarify the unit tests which have little need to concern themselves with rate.

File: py_zipkin/request_helpers.py (hunk 21 above)
Comment (adriancole) [reply py_zipkin/request_helpers.py]: sorry I'll be more concrete

parsed['sampled'] = true or false regardless of the shape of headers, unless the sampled value was absent.
parsed['debug'] = true only when 'd' in single format or iff 'X-B3-Flags: 1' in multi

basically the header extract function cleans the data into atoms like above.

File: py_zipkin/request_helpers.py (hunk 26 above)
Comment (adriancole) [reply py_zipkin/request_helpers.py]: I'll leave this point up to @drolando as it is more about maintenance than anything else, and I don't maintain this particular code.

**Comment by adriancole** [comment]
thanks for the work here and good luck!

**Comment by dbishop** [comment]
New commit up; I'll reiterate that based on how many edge cases got caught just in this clean-up, the level of testing here is quite appropriate.

File: py_zipkin/request_helpers.py (hunk 19 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Interesting.

Like you, I don't maintain this codebase either.  But this is a new helper-function.  In the past, anyone using `py_zipkin` had to write their own functions to do this.  So no existing users will be hampered by any lack of flexible composition here.

I think anyone needing that use-case could perform follow-on work here to add that complexity when it's necessary.

One of the things I liked about `py_zipkin` when looking for a python tracing library was how it did nearly all I wanted it to (subject to a couple open PRs, hehe), and not much more.

File: py_zipkin/request_helpers.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Thanks!  I think it's probably an "agree to disagree" situation on the level of pedantry here. :)

File: py_zipkin/request_helpers.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (dbishop) [reply py_zipkin/request_helpers.py]: I was indeed able to move the necessary (subject to our disagreement about whether this code should be flexibly composed in a more general way) logic for this elsewhere and get rid of the silly dict key.

File: py_zipkin/request_helpers.py (hunk 21 above)
Comment (dbishop) [reply py_zipkin/request_helpers.py]: Ah, okay, makes sense.  I think the new code does this.  In both multi and single-header cases, the sampling decision is normalized to `'0', '1', 'd', or None (defer)`

Comment (adriancole) [reply py_zipkin/request_helpers.py]: sound good

**Comment by dbishop** [comment]
I think this is cleanly rebased against black-formatted `master` branch.  Please let me know if you want me to do anything else on this PR.  I think the tests, though pretty long and exhaustive, did help guide implementation changes, refactorings, etc. and have great value for catching regressions.
